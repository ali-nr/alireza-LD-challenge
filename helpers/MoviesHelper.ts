import {MoviesData} from '../api/MoviesApi';
import {AppColors} from '../Colors';

export type Movie = {
  actors: string;
  id: string;
  poster: string;
  title: string;
  type: string;
  cinemaWorldPrice?: number;
  filmWorldPrice?: number;
};

// Assumption here is that both APIs return the same exact list of films. 
// Otherwise this function needs improvement 
export function movies(
  filmWorldData: MoviesData | undefined,
  cinemaWorldData: MoviesData | undefined,
): Movie[] | undefined {
  return filmWorldData?.Movies.map(filmWorldMovie => {
    const cinemaWorldMovies = cinemaWorldData?.Movies.find(cinemaWorldMovie => {
      // Needs improvements on probably check with IDs and not title 
      return cinemaWorldMovie.Title === filmWorldMovie.Title;
    });

    return {
      id: filmWorldMovie.ID,
      title: filmWorldMovie.Title,
      actors: filmWorldMovie.Actors,
      poster: filmWorldMovie.Poster,
      type: filmWorldMovie.Type,
      cinemaWorldPrice: cinemaWorldMovies?.Price,
      filmWorldPrice: filmWorldMovie.Price,
    } as Movie;
  });
}

export function priceBackground(
  firstPrice?: number,
  secondPrice?: number,
): string {
  if (firstPrice === undefined || secondPrice === undefined) {
    return AppColors.neutralPriceBackgroundColor;
  }
  return firstPrice <= secondPrice
    ? AppColors.cheapPriceBackground
    : AppColors.expensivePriceBackground;
}
