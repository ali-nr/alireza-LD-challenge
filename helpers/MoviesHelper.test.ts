import {cinemaWorldData, filmWorldData} from '../TestData';
import {movies, priceBackground} from './MoviesHelper';

describe('MoviesHelper', () => {
  it('Should return correct movies format', () => {
    const allMovies = movies(filmWorldData, cinemaWorldData);

    expect(allMovies).toMatchSnapshot();
  });

  it('Should return the correct color based on price', () => {
    const cheapPriceColor = priceBackground(33, 37);
    expect(cheapPriceColor).toEqual('#66bb6a');

    const expensivePriceColor = priceBackground(37, 33);
    expect(expensivePriceColor).toEqual('#ef5350');

    const neutralPriceColor = priceBackground(37, undefined);
    expect(neutralPriceColor).toEqual('#26c6da');
  });
});
