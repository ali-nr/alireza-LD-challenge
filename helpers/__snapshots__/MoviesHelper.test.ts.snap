// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MoviesHelper Should return correct movies format 1`] = `
Array [
  Object {
    "actors": "Harrison Ford, Mark Hamill, Carrie Fisher, Adam Driver",
    "cinemaWorldPrice": 24.7,
    "filmWorldPrice": 25,
    "id": "fw2488496",
    "poster": "https://m.media-amazon.com/images/M/MV5BOTAzODEzNDAzMl5BMl5BanBnXkFtZTgwMDU1MTgzNzE@._V1_SX300.jpg",
    "title": "Star Wars: Episode VII - The Force Awakens",
    "type": "movie",
  },
  Object {
    "actors": "Mark Hamill, Carrie Fisher, Adam Driver, Daisy Ridley",
    "cinemaWorldPrice": 24,
    "filmWorldPrice": 24.5,
    "id": "fw2527336",
    "poster": "https://m.media-amazon.com/images/M/MV5BMjQ1MzcxNjg4N15BMl5BanBnXkFtZTgwNzgwMjY4MzI@._V1_SX300.jpg",
    "title": "Star Wars: Episode VIII - The Last Jedi",
    "type": "movie",
  },
  Object {
    "actors": "Carrie Fisher, Mark Hamill, Adam Driver, Daisy Ridley",
    "cinemaWorldPrice": 23,
    "filmWorldPrice": 23.5,
    "id": "fw2527338",
    "poster": "https://m.media-amazon.com/images/M/MV5BMDljNTQ5ODItZmQwMy00M2ExLTljOTQtZTVjNGE2NTg0NGIxXkEyXkFqcGdeQXVyODkzNTgxMDg@._V1_SX300.jpg",
    "title": "Star Wars: Episode IX - The Rise of Skywalker",
    "type": "movie",
  },
  Object {
    "actors": "Felicity Jones, Diego Luna, Alan Tudyk, Donnie Yen",
    "cinemaWorldPrice": 25,
    "filmWorldPrice": 28,
    "id": "fw3748528",
    "poster": "https://m.media-amazon.com/images/M/MV5BMjEwMzMxODIzOV5BMl5BanBnXkFtZTgwNzg3OTAzMDI@._V1_SX300.jpg",
    "title": "Rogue One: A Star Wars Story",
    "type": "movie",
  },
  Object {
    "actors": "Alden Ehrenreich, Joonas Suotamo, Woody Harrelson, Emilia Clarke",
    "cinemaWorldPrice": 24.5,
    "filmWorldPrice": 24,
    "id": "fw3778644",
    "poster": "https://m.media-amazon.com/images/M/MV5BOTM2NTI3NTc3Nl5BMl5BanBnXkFtZTgwNzM1OTQyNTM@._V1_SX300.jpg",
    "title": "Solo: A Star Wars Story",
    "type": "movie",
  },
  Object {
    "actors": "Mark Hamill, Harrison Ford, Carrie Fisher, Peter Cushing",
    "cinemaWorldPrice": 25.5,
    "filmWorldPrice": 22.9,
    "id": "fw0076759",
    "poster": "https://m.media-amazon.com/images/M/MV5BNzVlY2MwMjktM2E4OS00Y2Y3LWE3ZjctYzhkZGM3YzA1ZWM2XkEyXkFqcGdeQXVyNzkwMjQ5NzM@._V1_SX300.jpg",
    "title": "Star Wars: Episode IV - A New Hope",
    "type": "movie",
  },
  Object {
    "actors": "Mark Hamill, Harrison Ford, Carrie Fisher, Billy Dee Williams",
    "cinemaWorldPrice": 23,
    "filmWorldPrice": 23.7,
    "id": "fw0080684",
    "poster": "https://m.media-amazon.com/images/M/MV5BYmU1NDRjNDgtMzhiMi00NjZmLTg5NGItZDNiZjU5NTU4OTE0XkEyXkFqcGdeQXVyNzkwMjQ5NzM@._V1_SX300.jpg",
    "title": "Star Wars: Episode V - The Empire Strikes Back",
    "type": "movie",
  },
  Object {
    "actors": "Mark Hamill, Harrison Ford, Carrie Fisher, Billy Dee Williams",
    "cinemaWorldPrice": 24.2,
    "filmWorldPrice": 22,
    "id": "fw0086190",
    "poster": "https://m.media-amazon.com/images/M/MV5BOWZlMjFiYzgtMTUzNC00Y2IzLTk1NTMtZmNhMTczNTk0ODk1XkEyXkFqcGdeQXVyNTAyODkwOQ@@._V1_SX300.jpg",
    "title": "Star Wars: Episode VI - Return of the Jedi",
    "type": "movie",
  },
  Object {
    "actors": "Liam Neeson, Ewan McGregor, Natalie Portman, Jake Lloyd",
    "cinemaWorldPrice": 26.4,
    "filmWorldPrice": 27.2,
    "id": "fw0120915",
    "poster": "https://m.media-amazon.com/images/M/MV5BYTRhNjcwNWQtMGJmMi00NmQyLWE2YzItODVmMTdjNWI0ZDA2XkEyXkFqcGdeQXVyNTAyODkwOQ@@._V1_SX300.jpg",
    "title": "Star Wars: Episode I - The Phantom Menace",
    "type": "movie",
  },
  Object {
    "actors": "Ewan McGregor, Natalie Portman, Hayden Christensen, Christopher Lee",
    "cinemaWorldPrice": 20.5,
    "filmWorldPrice": 19.9,
    "id": "fw0121765",
    "poster": "https://m.media-amazon.com/images/M/MV5BMDAzM2M0Y2UtZjRmZi00MzVlLTg4MjEtOTE3NzU5ZDVlMTU5XkEyXkFqcGdeQXVyNDUyOTg3Njg@._V1_SX300.jpg",
    "title": "Star Wars: Episode II - Attack of the Clones",
    "type": "movie",
  },
  Object {
    "actors": "Ewan McGregor, Natalie Portman, Hayden Christensen, Ian McDiarmid",
    "cinemaWorldPrice": 23,
    "filmWorldPrice": 22.4,
    "id": "fw0121766",
    "poster": "https://m.media-amazon.com/images/M/MV5BNTc4MTc3NTQ5OF5BMl5BanBnXkFtZTcwOTg0NjI4NA@@._V1_SX300.jpg",
    "title": "Star Wars: Episode III - Revenge of the Sith",
    "type": "movie",
  },
]
`;
